Layout: Add `preview` layoutFocus for `DesignPreview`

sirbrillig opened the pull request:
# Summary
Adds a new layoutFocus called `preview` which, when activated (`layoutFocus.set( 'preview' )`), shows a preview of the current site pretty much anywhere in Calypso.

![preview-layout-demo](https://cloud.githubusercontent.com/assets/2036909/14575403/96f4df48-032f-11e6-961b-ac64a159e395.gif)

# How it works
The `DesignPreview` component can render raw markup to a preview and then apply customizations to that preview. This PR adds a data layer for the preview markup and the customizations.

The raw markup is fetched from the REST API and saved as `previewMarkup` in the Redux state tree (one for each site). When it is rendered, the method `updatePreviewWithChanges()` in `lib/design-preview` will be called with the arguments:

1. The document object of the iframe.
2. The current value of the `customizations` object from the Redux state tree for this site.

The `customizations` object and the `previewMarkup` used by the preview are stored in the Redux state tree under the `preview` key for a given site. The state also keeps track of the "saved" status of the customizations and a history of all customization changes (enabling an "undo" feature).

This PR is essentially a newer version of several parts of #4140 and #4153 except without `DesignMenu`, the updaters, or the save-functions, which can be added separately.

# Testing
To activate it right now you can press the "customize" button in the sidebar, but that is only temporary and **should be removed before merging**. It would be nice if we could replace that with a command-line method of activating the layout. Not sure how to go about that...

sirbrillig commented:
Ping @mtias, @ehg, and @mcsf for comments

mcsf commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -935,6 +935,40 @@ Undocumented.prototype.setPrimaryDomain = function( siteId, domain, fn ) {
+		} )
+		.catch( reject );
+	} );
+};
```
Two notes:
- I think we should add one step of indentation when starting a line with `.then`, `.catch`, or any other property/method access, and that is AFAIK the convention in Calypso.
- I'd refactor things here, as the handling of promises is repeated verbatim in the function:

```js
const request = postData && Object.keys( postData ).length > 0
  ? this.wpcom.req.post( endpoint, query, { customized: postData } )
  : this.wpcom.req.get( endpoint, query );

request
  .then( response => { … } )
  .catch( reject );
```

As a side effect, it makes the function more self-explanatory.

sirbrillig replied:
Good points! I don't like long ternaries as I think they decrease readability, but I also think combining the requests is a good idea, so I compromised and made a shorter ternary with some extra `const`s.

mcsf commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -935,6 +935,33 @@ Undocumented.prototype.setPrimaryDomain = function( siteId, domain, fn ) {
+		const endpoint = `/sites/${siteId}/previews/mine`;
+		const query = { path: slug };
+		const isPreviewCustomized = ( postData && Object.keys( postData ).length > 0 );
+		const { post, get } = this.wpcom.req;
```
Interesting. I didn't expect `post` and `get` to be automatically bound to `req`.

sirbrillig replied:
Oh yeah, interesting indeed. I didn't expect that `get` or `post` would use `this`, but [there it is](https://github.com/Automattic/wpcom.js/blob/master/lib/util/request.js#L29). It looks like maybe binding to the right-hand-side object is part of object destructuring assignment.

mcsf replied:
It still makes me uncomfortable, though. :) And, actually, I'm not convinced that there's any magic binding happening: see [Babel output](https://babeljs.io/repl/#?evaluate=true&presets=es2015%2Creact%2Cstage-2&experimental=true&loose=false&spec=false&code=function%20bar()%20%7B%0A%20%20const%20%7B%20post%2C%20get%20%7D%20%3D%20this.wpcom.req%3B%0A%20%20post(%20foo%20)%3B%0A%7D).

I'd keep the method calls attached to the whole context tree (`this.wpcom.req.post( bla )`). Then again, I am not bothered by three-line ternaries. ;)

sirbrillig replied:
Fair enough. Restored the longer version.

mcsf commented on client/layout/index.jsx:
```diff
@@ -144,6 +145,17 @@ Layout = React.createClass( {
 	},
 
+	renderPreview() {
+		if ( config.isEnabled( 'preview-layout' ) && this.props.section.group === 'sites' ) {
```
Was going to ask whether it made more sense to move the first condition (config check) into `DesignPreview` proper, letting the component itself decide whether it's enabled.

That'd would also make the remainder condition (check that group === sites) smaller and maybe make it more comfortable to ditch `renderPreview` altogether and have the thing inlined in the `render`, possibly removing indirection. #nit-picks

sirbrillig replied:
An interesting idea! I am pondering it. It seems like there are trade-offs here since if we move the condition into `DesignPreview` then in a way it tightly binds that class to that config check. OTOH, it may be a temporary config and it's unlikely we'd use the object anywhere else in Calypso...

mcsf commented on client/layout/sidebar/item.jsx:
```diff
@@ -19,6 +19,7 @@ export default React.createClass( {
 		className: React.PropTypes.string,
 		link: React.PropTypes.string.isRequired,
 		buttonLink: React.PropTypes.string,
+		onButtonClick: React.PropTypes.func,
```
Maybe a note that this overrides the default click action?

sirbrillig replied:
That's temporary and has been deprecated by conflicts anyway. Removed.

mcsf commented on client/my-sites/design-preview/README.md:
```diff
@@ -0,0 +1,12 @@
+# DesignPreview
+
+A wrapper for the always-available instance of `WebPreview` that's activated when `layoutFocus` is set to `preview`.
```
❤️ proper hyphenation

mcsf commented on client/my-sites/design-preview/README.md:
```diff
@@ -0,0 +1,12 @@
+
+A wrapper for the always-available instance of `WebPreview` that's activated when `layoutFocus` is set to `preview`.
+
+The `DesignPreview` component can render raw markup to a preview and then apply customizations to that preview. This PR adds a data layer for the preview markup and the customizations.
```
Suggestion:

```diff
-The `DesignPreview` component can render raw markup to a preview and then apply
+The `DesignPreview` component can render raw markup to a preview (see `WebPreview`) and then apply
```

mcsf replied:
Actually, scratch that, since `WebPreview` is mentioned right in the previous sentence.

mcsf commented on client/my-sites/design-preview/index.js:
```diff
@@ -0,0 +1,171 @@
+		}
+		// If the customizations have been removed, restore the original markup
+		if ( this.props.previewMarkup && this.props.customizations && this.props.previewMarkup === prevProps.previewMarkup && prevProps.customizations ) {
+			if ( Object.keys( this.props.customizations ).length === 0 && Object.keys( prevProps.customizations ).length > 0 ) {
```
The nested `if` can be merged into the parent.

sirbrillig replied:
👍

mcsf commented on client/my-sites/design-preview/index.js:
```diff
@@ -0,0 +1,171 @@
+	loadPreviewWithProps( props ) {
+		if ( ! props ) {
+			props = this.props;
+		}
```
`loadPreviewWithProps` is never called with `props`. I take it it's a leftover detail?

sirbrillig replied:
Huh. Nice catch! I'm sure that was important at some point.

lsinger commented:
This looks great! Two things I noticed after playing around a bit:

- When I change my site (in this case, I changed my header image) the preview doesn't refresh, even if I reload Calypso.dev. Should it notice changes by itself? Should I have a way to refresh? 

- When I open the preview, it opens in tablet size by default. I prefer desktop size, so I set it to that. I close the preview. I open the preview again. It's now in desktop size, great--but apparently the CSS stops working. Can anyone reproduce or is it just me?

sirbrillig commented:
> When I change my site (in this case, I changed my header image) the preview doesn't refresh, even if I reload Calypso.dev. Should it notice changes by itself? Should I have a way to refresh?

I think that was a bug having to do with the site ID not being available when the component loads. I've just fixed it in 9153d4e. Let me know if that solves the issue for you.

> When I open the preview, it opens in tablet size by default. I prefer desktop size, so I set it to that. I close the preview. I open the preview again. It's now in desktop size, great--but apparently the CSS stops working. Can anyone reproduce or is it just me?

@lsinger Interesting! I wasn't able to reproduce but it may be something about your particular site. Can you share the site with me? (we can do this via Slack.)

sirbrillig commented:
Rebased to remove conflicts. Sorry to anyone who had this checked out.

lsinger commented:
> I think that was a bug having to do with the site ID not being available when the component loads. I've just fixed it in 9153d4e. Let me know if that solves the issue for you.

Yup, reloading Calypso does the trick now. 👍 

Would it make sense to (while the preview is open) regularly poll the API whether anything about the site has changed? 

> Interesting! I wasn't able to reproduce but it may be something about your particular site. Can you share the site with me? (we can do this via Slack.)

Sent it to you in Slack. See also the following font differences. 

**Actual site:**

<img width="609" alt="live" src="https://cloud.githubusercontent.com/assets/23619/14980180/4778ba7e-1127-11e6-8466-d5af3b94ca7a.png">

**Preview:**

<img width="559" alt="preview" src="https://cloud.githubusercontent.com/assets/23619/14980191/5506ef26-1127-11e6-9211-673d85aac285.png">

sirbrillig commented:
> Would it make sense to (while the preview is open) regularly poll the API whether anything about the site has changed?

Hm... Probably? OTOH polling is expensive, especially here because the preview is large. I'm open to discussion but I think it might be best to put that off as a feature for later (perhaps with a more efficient mechanism to query for a changed site).

lsinger commented:
> it might be best to put that off as a feature for later 

That's what I thought, just wanted to make sure I'd mentioned it. :)

ehg commented on client/state/preview/reducer.js:
```diff
@@ -0,0 +1,69 @@
+/**
+ * External dependencies
+ */
+import assign from 'lodash/assign';
```
Any reason not to use `Object.assign` instead of lodash?

sirbrillig replied:
Because, as I found out relatively recently, `Object.assign` is ES6 and (at least last time I checked) we don't support polyfills for prototype methods in Calypso. So if I use `Object.assign()` then it will only work in browsers that support that feature.

ehg replied:
It isn't a prototype method :) Ref: https://github.com/Automattic/wp-calypso/issues/2360#issuecomment-171296212

sirbrillig replied:
> Half true. We're not actually using babel/polyfill, we're using the babel/runtime with webpack, which only polyfills non-prototype things. So we get Map and Set and Object.assign, but not things on String.prototype like startsWith.

Ohhhhh... TIL.

ehg commented on client/state/preview/reducer.js:
```diff
@@ -0,0 +1,69 @@
+/**
+ * Internal dependencies
+ */
+import * as ActionTypes from 'state/action-types';
```
One thing to note is that when we move to a tree-shaking build tool, e.g. webpack2, `import *` won't be optimised. This doesn't matter for now, and won't matter for this particular import — it's used globally and is tiny — but it's something to bear in mind anyway.

ehg commented:
I think this is looking like a pretty solid start. Nice job. No doubt we'll find bugs/missing features when we use it in anger, but for me it makes sense to remove the 'Customize' hijack and merge this.

@lsinger @mcsf what do you think?

ehg commented on client/my-sites/design-preview/index.js:
```diff
@@ -63,6 +63,12 @@ const DesignPreview = React.createClass( {
+		// Refresh the preview when it is being shown (since this component is
+		// always present but not always visible, this is similar to loading the
+		// preview when mounting).
+		if ( this.props.showPreview && ! prevProps.showPreview ) {
```
Nifty iterim solution!

lsinger commented:
> I think this is looking like a pretty solid start. Nice job. No doubt we'll find bugs/missing features when we use it in anger, but for me it makes sense to remove the 'Customize' hijack and merge this now. 

The incorrect fonts still worry me--also could point to a deeper problem.

ehg commented:
> The incorrect fonts still worry me--also could point to a deeper problem.

Hmm, interestingly the webfonts work for me in Firefox, but not in Chrome (consistently — I saw a brief flash). Taking a deeper look.

ehg commented:
> The incorrect fonts still worry me--also could point to a deeper problem.

Updating https://github.com/typekit/webfontloader on my WPCOM sandbox to the current version fixes this for Google Fonts in Chrome. TypeKit fonts will only work on wordpress.com domains.

ehg commented:
I just got a 'your stats are booming!' notification. We shouldn't be bumping stats for previews…

sirbrillig commented:
> I just got a 'your stats are booming!' notification. We shouldn't be bumping stats for previews

Ah, yes. 😩 I had forgotten about this. It's an issue with the Previews endpoint. @mattwiebe might have some insight.

ehg commented:
Seeing as the webfonts and stat bumping are issues independent of this PR, I think we should merge this, and track them elsewhere.

sirbrillig commented:
I'll rebase to remove the sidebar hack and mark as Ready for Review

sirbrillig commented:
Rebased. Sidebar "customize" hook is still available in 672cc9a4c10be24be8f90fee773dbee222596177 if needed for testing.

ehg commented:
LGTM. :shipit: and onwards to https://github.com/Automattic/wp-calypso/pull/5117 :)

aduth commented on client/my-sites/design-preview/index.js:
```diff
@@ -0,0 +1,182 @@
+	loadPreview() {
+		if ( this.props.selectedSiteId && this.props.actions.fetchPreviewMarkup ) {
+			debug( 'loading preview with customizations', this.props.customizations );
+			this.props.actions.fetchPreviewMarkup( this.props.selectedSiteId, '', this.props.customizations );
```
Do we want to be triggering a network request for preview whenever navigating My Sites? At least, this is what I'm observing currently (assuming component is always mounted in the background?)

sirbrillig replied:
Hm. That doesn't sound right. I think my intent was that the preview request only load when either: 

1. The my-sites area for a specific site first loads.
2. The preview is made visible (in case there are changes).
3. When customizations had been applied but have since been removed.

aduth replied:
@sirbrillig Unfortunately that doesn't appear to be the case. Here's an example refreshing the settings page, showing the network request on page load:

![request](https://cloud.githubusercontent.com/assets/1779930/15431540/5ff26b14-1e78-11e6-946f-5a3d1a679064.gif)

sirbrillig replied:
Yeah, no good. We may want to speed up the merge of parts of #5117 which disable the preview endpoint.

ehg replied:
> 1. The my-sites area for a specific site first loads.
2. The preview is made visible (in case there are changes).
3. When customizations had been applied but have since been removed.

> an example refreshing the settings page, showing the network request on page load…

Doesn't this conform to the expected behaviour? I believe we want to load the preview on the initial page load, and for every site switched. We could switch it to fetch on demand, but we'd lose the 'instant' factor, which the editor preview has.

aduth replied:
>Doesn't this conform to the expected behaviour? I believe we want to load the preview on the initial page load, and for every site switched. We could switch it to fetch on demand, but we'd lose the 'instant' factor, which the editor preview has.

Eh, that feels a bit heavy. The likelihood that my session using My Sites will lead to previewing customizations is much much lower than the likelihood I'd preview a post I'm editing.

aduth replied:
Not sure what the behavior is here, but we're also not doing background loading for mobile devices in the editor.

I'm not a fan of the editor background loading FWIW 😄  I wonder if there's a middle ground where we interpret "intent" (e.g. hover action).

sirbrillig replied:
I'd like to get @mtias input also because the background loading (once for each site, when it first loads) was originally his idea.
